### PR TITLE
Servlet lesson 18-19

### PR DESCRIPTION
**Pull request by AndreiChernukho**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  <artifactId>untitled</artifactId>
+  <version>1.0-SNAPSHOT</version>
+  <packaging>war</packaging>
+
+  <name>untitled Maven Webapp</name>
+  <url>http://www.example.com</url>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>1.7</maven.compiler.source>
```
Comment (gr00shik) [inline pom.xml]: 1.8 at least

File: src/main/java/CarService.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import entity.Car;
+
+import java.util.List;
+
+public interface CarService {
+    List<Car> getAll();
+
+    Car get(int id);
+
+    void create(Long id, String carModel);
```
Comment (gr00shik) [inline src/main/java/CarService.java]: create(Car car)

File: src/main/java/CarService.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+    void update(Long id, String carModel);
```
Comment (gr00shik) [inline src/main/java/CarService.java]: update(Car car)

File: src/main/java/CarServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import entity.Car;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CarServiceImpl implements CarService {
+    private static final Map<Long, Car> CAR_MAP = new HashMap<>();
```
Comment (gr00shik) [inline src/main/java/CarServiceImpl.java:9]: you can use set instead of map

File: src/main/java/CarServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    public void create(Long id, String carModel) {
+        Car car = new Car(id, carModel, LocalDateTime.now());
+        CAR_MAP.put(id, car);
+    }
+
+    @Override
+    public void update(Long id, String carModel) {
```
Comment (gr00shik) [inline src/main/java/CarServiceImpl.java]: what difference between create and update

File: src/main/java/CarServlet.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+@WebServlet(value = "/car")
+public class CarServlet extends HttpServlet {
+
+    private static final String CAR_ID_PARAMETER_NAME = "id";
+    private static final String CAR_MODEL_PARAMETER_NAME = "carModel";
+    private static final String TIME_COOKIE_NAME = "time";
+    private static final String CARS_ATTRIBUTE_NAME = "cars";
+    private static final String CARS_PAGE_NAME = "cars.jsp";
+
+    private final CarService carService = new CarServiceImpl();
```
Comment (gr00shik) [inline src/main/java/CarServlet.java]: put this logic to constructor

File: src/main/java/TestFilter.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.io.IOException;
+
+@WebFilter("/cars")
+public class TestFilter implements Filter {
+    private static final String SPECIAL_PARAMETER_NAME = "isValid";
+    private static final String ALLOWED_REQUEST_TYPE = "GET";
+    private static final String ERROR_PAGE_PATH = "/main/webapp/WEB-INF/error.jsp";
+
+    @Override
+    public void init(FilterConfig filterConfig) {
```
Comment (gr00shik) [inline src/main/java/TestFilter.java:13]: just remove it

File: src/main/java/TestFilter.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+        if (isValid || httpServletRequest.getMethod().equals(ALLOWED_REQUEST_TYPE)) {
+            chain.doFilter(request, response);
+        } else {
+            request.getRequestDispatcher(ERROR_PAGE_PATH).forward(request, response);
+        }
+    }
+
+    @Override
+    public void destroy() {
```
Comment (gr00shik) [inline src/main/java/TestFilter.java:30]: just remove it

File: src/main/webapp/cars.jsp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+    <title>Добро пожаловать, JSP!</title>
+</head>
+<body>
+<table border="1">
+    <tr>
+        <td>Car model:</td>
+        <td>Creation date:</td>
+    </tr>
+    <% List<Car> cars = (List<Car>) request.getAttribute("cars");
```
Comment (gr00shik) [inline src/main/webapp/cars.jsp:18]: jstl will be better

File: src/main/webapp/cars.jsp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    <tr>
+        <td><%= cars.get(i).getCarModel()%>
+        </td>
+        <td><%= cars.get(i).getCreationDate()%>
+        </td>
+    </tr>
+    <% } %>
+</table>
+<%
+    Cookie[] cookies = request.getCookies();
```
Comment (gr00shik) [inline src/main/webapp/cars.jsp:29]: use expression language for it to avoid java and difficult logic

**Review by gr00shik — COMMENTED** [review]
ok

File: src/main/java/TestFilter.java (hunk 7 above)
Comment (AndreiChernukho) [reply src/main/java/TestFilter.java:13]: Filter is interface, we need to reliase all methods.

File: src/main/java/TestFilter.java (hunk 8 above)
Comment (AndreiChernukho) [reply src/main/java/TestFilter.java:30]: Filter is interface, we need to reliase all methods.

File: src/main/java/CarService.java (hunk 2 above)
Comment (AndreiChernukho) [reply src/main/java/CarService.java]: fix

File: pom.xml (hunk 1 above)
Comment (AndreiChernukho) [reply pom.xml]: fix

File: src/main/java/CarService.java (hunk 3 above)
Comment (AndreiChernukho) [reply src/main/java/CarService.java]: fix

File: src/main/java/CarServiceImpl.java (hunk 5 above)
Comment (AndreiChernukho) [reply src/main/java/CarServiceImpl.java]: fix

**Pull request merged**